pkg/nsx/services/vpc/vpc.go

```diff
@@ -470,3 +521,232 @@ func (s *VPCService) Cleanup() error {
 
 	return nil
 }
+
+func (service *VPCService) needUpdateRule(rule *model.Rule, externalCIDR []string) bool {
+	des := rule.DestinationGroups
+	currentDesSet := sets.Set[string]{}
+	for _, group := range des {
+		currentDesSet.Insert(group)
+	}
+	if len(externalCIDR) != len(currentDesSet) {
+		return true
+	}
+	for _, newGroup := range externalCIDR {
+		if !currentDesSet.Has(newGroup) {
+			return true
+		}
+	}
+	return false
+}
+
+func (service *VPCService) getIpblockCidr(cidrID []string) (result []string, err error) {
+	for _, cidr := range cidrID {
+		ipblock := service.PubIpblockStore.GetByKey(cidr)
+		if ipblock == nil {
+			// in case VPC using the new ipblock, search the ipblock from nsxt
+			// return error, and retry next time when the ipblock is synced into store
+			err = errors.New("ipblock not found")
+			log.Error(err, "failed to get public ipblock", "path", cidr)
+			log.Info("query public ipblock from nsxt ")
+			query := fmt.Sprintf("%s:%s AND visibility:EXTERNAL", common.ResourceType, common.ResourceTypeIPBlock)
+			count, searcherr := service.SearchResource(common.ResourceTypeIPBlock, query, service.PubIpblockStore, nil)
+			if searcherr != nil {
+				log.Error(searcherr, "failed to query public ipblock", "query", query)
+			} else {
+				log.Info("query public ipblock", "count", count)
+			}
+			return
+		} else {
+			result = append(result, *ipblock.Cidr)
+		}
+	}
+	return
+}
+
+func (service *VPCService) CreateOrUpdateAVIRule(vpc *model.Vpc, namespace string) error {
+	if !enableAviAllowRule {
+		return nil
+	}
+	vpcInfo, err := common.ParseVPCResourcePath(*vpc.Path)
+	orgId := vpcInfo.OrgID
+	projectId := vpcInfo.ProjectID
+	ruleId := AllowAviRuleId
+	groupId := AllAviSEGroupId
+	spId := AviSecurityPolicyId
+
+	if !service.checkAVISecurityPolicyExist(orgId, projectId, *vpc.Id, spId) {
+		return errors.New("avi security policy not found")
+	}
+	allowrule, err := service.getAVIAllowRule(orgId, projectId, *vpc.Id, spId, ruleId)
+	externalCIDR, err := service.getIpblockCidr(vpc.ExternalIpv4Blocks)
+	if err != nil {
+		return err
+	} else {
+		log.Info("avi rule get external cidr", "cidr", externalCIDR)
+	}
+
+	if allowrule != nil {
+		if !service.needUpdateRule(allowrule, externalCIDR) {
+			log.Info("avi rule not changed, skip update avi rule")
+			return nil
+		} else {
+			log.Info("avi rule changed", "previous", allowrule.DestinationGroups, "current", externalCIDR)
```

Comment:
As we define the group with fixed name, printing previous group does not have any useful info, right?